### PR TITLE
Improve `distinct` by using `cuco::static_map::retrieve_all`

### LINKED ISSUE #10909: [BUG] `cudf::distinct` implementation missed stream parameter when calling to `key_map.contains`
This is a small issue. As the title says: `cudf::distinct` implementation (in file `distinct.cu` missed stream parameter when calling to `key_map.contains`. We should correct that whenever possible.

### PR DESCRIPTION
**Pull request by PointKernel**
Closes #10909

This PR was intended to fix a bug in the `distinct` implementation where the stream parameter was not passed when invoking `static_map::contains`. During the work, @ttnghia Pointed out that the `contains` + `thrust::copy_if` logic can be simplified by using `static_map::retrieve_all`. Finally, the PR fetches a newer version of `cuco` to utilize `retrieve_all` and fixes a bug in unit tests where results should be sorted before comparison.

### CODE REVIEW

**Review by vuule — APPROVED** [review]

**Review by vyasr — APPROVED** [review]

File: cpp/src/stream_compaction/distinct.cu [hunk 1]
```diff
@@ -82,7 +82,12 @@ std::unique_ptr<table> distinct(table_view const& input,
   auto counting_iter = thrust::make_counting_iterator<size_type>(0);
   rmm::device_uvector<bool> index_exists_in_map(num_rows, stream, mr);
   // enumerate all indices to check if they are present in the map.
-  key_map.contains(counting_iter, counting_iter + num_rows, index_exists_in_map.begin(), hash_key);
+  key_map.contains(counting_iter,
+                   counting_iter + num_rows,
+                   index_exists_in_map.begin(),
+                   hash_key,
+                   thrust::equal_to<size_type>{},
+                   stream.value());
```
Comment (ttnghia) [inline cpp/src/stream_compaction/distinct.cu:90]: Do we have a `get_keys` method to just extract a list of all keys from the map?

Comment (ttnghia) [reply cpp/src/stream_compaction/distinct.cu:90]: I found `retrieve_all` but not sure what is it for?

Comment (PointKernel) [reply cpp/src/stream_compaction/distinct.cu:90]: Good point! `retrieve_all` it is.

File: cpp/src/stream_compaction/distinct.cu [hunk 2]
```diff
@@ -79,23 +80,13 @@ std::unique_ptr<table> distinct(table_view const& input,
...
   auto mutable_view = mutable_column_device_view::create(*distinct_indices, stream);
-  thrust::copy_if(rmm::exec_policy(stream),
-                  counting_iter,
-                  counting_iter + num_rows,
-                  index_exists_in_map.begin(),
-                  mutable_view->begin<size_type>(),
-                  thrust::identity<bool>{});
+  // write distinct indices to a numeric column
+  key_map.retrieve_all(
+    mutable_view->begin<size_type>(), thrust::make_discard_iterator(), stream.value());
```
Comment (ttnghia) [inline cpp/src/stream_compaction/distinct.cu:89]:
```suggestion
  // write distinct indices to a numeric column
  key_map.retrieve_all(
    distinct_indices->mutable_view().template begin<size_type>(), thrust::make_discard_iterator(), stream.value());
```

**Comment by jrhemstad** [comment]
rerun tests

File: cpp/src/stream_compaction/distinct.cu (hunk 2 above)
Comment (jrhemstad) [reply cpp/src/stream_compaction/distinct.cu:89]: A template disambiguator is only needed when the symbol in question is a [dependent name](https://en.cppreference.com/w/cpp/language/dependent_name) (which on a brief look it appears it is not). 

If this code compiled w/o the disambiguation, then it isn't necessary.

Comment (jrhemstad) [reply cpp/src/stream_compaction/distinct.cu:89]: I do agree with eliminating the need for a `mutable_column_device_view` here.

**Comment by PointKernel** [comment]
Seeing performance regression when using `retrieve_all`:

- `static_map::contains` + `thrust::copy_if`:

| Type | NumRows  | Samples |  CPU Time  | Noise |  GPU Time  | Noise |
|------|----------|---------|------------|-------|------------|-------|
|  I32 |    10000 |   4640x | 113.619 us | 6.63% | 108.028 us | 4.10% |
|  I32 |   100000 |   3552x | 146.936 us | 5.29% | 141.284 us | 3.37% |
|  I32 |  1000000 |   1168x | 437.271 us | 1.85% | 431.588 us | 1.27% |
|  I32 | 10000000 |    142x |   3.534 ms | 0.44% |   3.529 ms | 0.41% |

- `static_map::retrieve_all`:

| Type | NumRows  | Samples |  CPU Time  | Noise  |  GPU Time  | Noise  |
|------|----------|---------|------------|--------|------------|--------|
|  I32 |    10000 |   1424x | 463.748 us |  6.15% | 457.890 us |  6.01% |
|  I32 |   100000 |   1408x | 500.350 us |  6.76% | 494.447 us |  6.64% |
|  I32 |  1000000 |   3152x | 760.985 us | 26.55% | 755.105 us | 26.54% |
|  I32 | 10000000 |    624x |   3.321 ms |  0.96% |   3.315 ms |  0.94% |

Will look into profiling tomorrow.

**Comment by ttnghia** [comment]
> Seeing performance regression when using `retrieve_all`:
Did you get these numbers by pulling the same `GIT_TAG` for cuco?

**Comment by jrhemstad** [comment]
The perf issue sounds similar to what @cheinger did in https://github.com/rapidsai/cudf/pull/10511 

He found explicitly materializing the gather mask improved perf because putting the device-side `contains` call into the `gather` increased register pressure too much. 

How is the `retrieve_all` implemented?

**Comment by PointKernel** [comment]
> The perf issue sounds similar to what @cheinger did in #10511 

Agreed. 

> How is the `retrieve_all` implemented?

`retrieve_all` is using `thrust::copy_if` + transform iterator to generate the output since it requires a device functor to transform the content of a hash map slot to `thrust::tuple`. Is it worth trying to write our own kernel in cuco instead of using `thrust::copy_if`? I don't see another way to get rid of this register usage issue otherwise.

**Comment by jrhemstad** [comment]
> Is it worth trying to write our own kernel in cuco instead of using `thrust::copy_if`

Doubtful. I don't see why there would be an increased register usage problem in this situation. The problem from https://github.com/rapidsai/cudf/pull/10511 was that we were using a `contains` function as the predicate for a `copy_if` which is a complex function that uses a lot of registers.

For `retrieve_all` it should be a incredibly simple predicate function of just checking if the slot is empty. So I think something else is going on.

**Comment by PointKernel** [comment]
Updates: the performance regression comes from a bug in the current `static_map::retrieve_all`. Created https://github.com/NVIDIA/cuCollections/pull/169 to fix it.

**Comment by PointKernel** [comment]
Updates: the performance regression comes from a bug in the current `static_map::retrieve_all`. Created https://github.com/NVIDIA/cuCollections/pull/169 to fix it.

**Comment by PointKernel** [comment]
Merging this PR since the performance regression issue will be resolved after merging #10983

**Comment by PointKernel** [comment]
@gpucibot merge

**Pull request merged**